File: Cpp/BaekJoon/Gold/G1_1300.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

long long N, k;

bool check(long long mid) {
  long long cnt = 0;

  for (int i = 1; i <= N; i++) {
    cnt += min(N, mid / i);
  }

  if (cnt >= k) return true;

  return false;
}

int main() {
  cin >> N >> k;

  long long lo = 1;
  long long hi = N * N;

  long long ans = 0;

  while (lo <= hi) {
    long long mid = (lo + hi) / 2;

    if (check(mid)) {
      ans = mid;
      hi = mid - 1;
    } else {
      lo = mid + 1;
    }
  }

  cout << ans;

  return 0;
}

File: Java/BaekJoon/Gold/G1_1300.java
package BaekJoon.Gold;

import java.util.Scanner;

public class G1_1300 {
    static boolean check(long N, long k, long x) {
        long cnt = 0;

        for (int i = 1; i <= N; i++) {
            cnt += Math.min(N, x / i);
        }

        if (cnt >= k) return true;

        return false;
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        long N = sc.nextLong();
        long k = sc.nextLong();

        long lo = 1;
        long hi = N * N;

        long ans = 0;

        while(lo <= hi) {
            long mid = (lo + hi) / 2;

            if (check(N, k, mid)) {
                ans = mid;
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        }

        System.out.println(ans);
        
        sc.close();
    }
}
